dev/buildtool/image_commands.py

```diff
@@ -18,91 +18,155 @@
+  def _do_determine_source_repositories(self):
+    """Implements RepositoryCommandProcessor interface."""
+    # These arent actually used, just the names.
+    repositories = [self.source_code_manager.make_repository_spec(name)
+                    for name in SPINNAKER_RUNNABLE_REPOSITORY_NAMES]
+    repositories.extend([
+        GitRepositorySpec(name) for name in EXTRA_REPO_NAMES])
+    return repositories
+
   def __init__(self, factory, options, **kwargs):
-    ensure_options_set(
+    check_options_set(
         options,
         ['build_gce_service_account',
          'build_gce_project',
          'publish_gce_image_project'])
 
-    # See class note as to why these are "repositories"
-    source_repositories = kwargs.pop(
-        'source_repositories', ALL_IMAGE_REPOSITORIES)
-    super(BuildGceComponentImages, self).__init__(
-        factory, options, source_repositories=source_repositories, **kwargs)
+    options.github_disable_upstream_push = True
+    super(BuildGceComponentImages, self).__init__(factory, options, **kwargs)
 
-  def __determine_repo_install_args(self):
+  def __determine_repo_install_args(self, repository):
     """Determine --spinnaker_dev-github_[owner|user] args for install script."""
     options = self.options
     branch = options.git_branch
     owner = ('spinnaker'
-             if options.github_user in ('default', 'upstream')
-             else options.github_user)
+             if options.github_owner in ('default', 'upstream')
+             else options.github_owner)
     git_dir = os.path.dirname(__file__)
     if not branch:
       branch = GitRunner(options).query_local_repository_branch(git_dir)
     if not owner:
-      url = (GitRunner(options)
-             .determine_remote_git_repository(git_dir)
-             .url)
-      owner = re.search('github.com/(.+)/spinnaker', url).group(1)
+      url = repository.origin
+      match = re.search('github.com/([^/]+)/', url)
+      if not match:
+        raise_and_log_error(
+            UnexpectedError('Cannot determine owner from url=%s' % url,
+                            cause='BadUrl'))
+      owner = match.group(1)
     return [
         '--spinnaker_dev_github_owner', owner,
         '--spinnaker_dev_github_branch', branch
     ]
 
+  def have_image(self, repository):
+    """Determine if we already have an image for the repository or not."""
+    bom = self.source_code_manager.bom
+    dependencies = bom['dependencies']
+    services = bom['services']
+    service_name = BomSourceCodeManager.to_service_name(repository)
+    if service_name in dependencies:
+      build_version = dependencies[service_name]['version']
+    else:
+      build_version = services[service_name]['version']
+
+    options = self.options
+    image_name = 'spinnaker-{repo}-{version}'.format(
+        repo=repository.name,
+        version=build_version.replace('.', '-').replace(':', '-'))
+    lookup_command = ['gcloud', '--account', options.build_gce_service_account,
```

Comment:
Are the `gcloud` resource lookup callsites enough of a pattern to warrant a helper?

Reply:
what are you thinking it would  look like?

Reply:
I was thinking maybe a subfunction taking a lookup command list and the artifact (from Spinnaker's vantage), doing the lookup, and then emitting the labels. Something like that. It might not fit the pattern exactly though.